geomstats/geometry/hyperbolic_space.py

```diff
@@ -321,42 +358,90 @@ def log(self, point, base_point):
 
-        coef_1 = gs.zeros_like(angle)
-        coef_2 = gs.zeros_like(angle)
+            coef_1 += mask_else_float * (angle / gs.sinh(angle))
+            coef_2 += mask_else_float * (angle / gs.tanh(angle))
 
-        coef_1 += mask_0_float * (
-                  1. + INV_SINH_TAYLOR_COEFFS[1] * angle ** 2
-                  + INV_SINH_TAYLOR_COEFFS[3] * angle ** 4
-                  + INV_SINH_TAYLOR_COEFFS[5] * angle ** 6
-                  + INV_SINH_TAYLOR_COEFFS[7] * angle ** 8)
-        coef_2 += mask_0_float * (
-                  1. + INV_TANH_TAYLOR_COEFFS[1] * angle ** 2
-                  + INV_TANH_TAYLOR_COEFFS[3] * angle ** 4
-                  + INV_TANH_TAYLOR_COEFFS[5] * angle ** 6
-                  + INV_TANH_TAYLOR_COEFFS[7] * angle ** 8)
+            log = (gs.einsum('ni,nj->nj', coef_1, point) -
+                   gs.einsum('ni,nj->nj', coef_2, base_point))
+            return log
 
-        # This avoids dividing by 0.
-        angle += mask_0_float * 1.
+        if self.point_type == 'poincare':
+            add_base_point = self.mobius_add(-base_point, point)
 
-        coef_1 += mask_else_float * (angle / gs.sinh(angle))
-        coef_2 += mask_else_float * (angle / gs.tanh(angle))
+            norm_add = add_base_point.norm(2,
+                                           -1, keepdim=True).expand_as(
+                add_base_point)
 
-        log = (gs.einsum('ni,nj->nj', coef_1, point)
-               - gs.einsum('ni,nj->nj', coef_2, base_point))
-        return log
+            norm_base_point = base_point.norm(2,
+                                              -1, keepdim=True).expand_as(
+                add_base_point)
+
+            res = (1 - norm_base_point ** 2) * \
+                  ((gs.arc_tanh(norm_add))) * (add_base_point / norm_add)
+
+            res[norm_add == 0] = 0
+
+            return res
+
+    def mobius_add(self, point_a, point_b):
+        """
+                Mobius addition operation that is necessary operation
+                to compute the log and exp using the 'poincare'
+                representation set as point_type.
+
+        Parameters
+        ----------
+        point_a : array-like, shape=[n_samples, dimension + 1]
+                              or shape=[1, dimension + 1]
+        point_b : array-like, shape=[n_samples, dimension + 1]
+                              or shape=[1, dimension + 1]
+
+        Returns
+        -------
+        mobius_add : array-like, shape=[n_samples, 1]
+                           or shape=[1, 1]
+        """
+        norm_point_a = gs.sum(point_a ** 2, dim=-1,
+                              keepdim=True).expand_as(point_a)
+        norm_point_b = gs.sum(point_b ** 2, dim=-1,
+                              keepdim=True).expand_as(point_a)
+        sum_prod_a_b = (point_a * point_b).sum(-1,
+                                               keepdim=True).expand_as(point_a)
+        return ((1 + 2 * sum_prod_a_b + norm_point_b) * point_a +
```

Comment:
This expression is too complicated: break it down in several lines for readability